blackjax/tempered_smc.py

```diff
@@ -0,0 +1,116 @@
+"""Public API for the Tempered SMC Kernel"""
+from typing import Callable, Dict, List, NamedTuple, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+import blackjax.inference.base as base
+from blackjax.hmc import HMCParameters
+
+Array = Union[np.ndarray, jnp.DeviceArray]
+PyTree = Union[Dict, List, Tuple]
+
+__all__ = ["new_state", "kernel"]
+
+
+class SMCState(NamedTuple):
+    """ Current state for the tempered SMC algorithm. This encodes the current set of sampled particles
+    as well as the corresponding tempering parameter :math:`\lambda`
+    """
+    particles: PyTree[Array]
+    lamda: float
+
+
+class SMCInfo(NamedTuple):
+    """Additional information on the tempered SMC step.
+
+    weights: jnp.ndarray
+        The weights after the MCMC pass.
+    """
+    weights: jnp.ndarray
+
+
+new_state = base.new_hmc_state
+
+
+def _normalize(log_weights):
+    """ Normalize log-weights into weights
+    """
+    w = jnp.exp(log_weights - jnp.max(log_weights))
+    return w / w.sum()
+
+
+def kernel(potential_fn: Callable, mcmc_kernel: Callable, new_mcmc_state: Callable, resampling_method: Callable,
+           mcmc_parameters: HMCParameters, mcmc_steps=10):
+    """Build a Tempered SMC step.
+
+    Parameters
+    ----------
+    potential_fn: Callable
+        A function that returns the potential energy of a chain at a given position.
+    mcmc_kernel: Callable
+        A callable function that creates a kernel from `potential_function` and parameters `mcmc_parameters`
+    new_mcmc_state: Callable
+        How to create a new MCMC state for the MCMC routine to start
+    resampling_method: Callable
+        A random function that resamples generated particles based of weights
+    mcmc_parameters: HMCParameters
+        The parameters to be fed to the mcmc kernel
+    mcmc_steps: int
+        Number of MCMC steps to run per SMC step
+
+    Returns
+    -------
+    A callable that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+    """
+
+    def one_step(rng_key: jnp.ndarray, state: SMCState, delta: float) -> Tuple[SMCState, SMCInfo]:
+        """
+
+        Parameters
+        ----------
+        rng_key: DeviceArray[int],
+            JAX PRNGKey for randomness
+        state: SMCState
+            Current state of the tempered SMC algorithm
+        delta: float
+            Increment for `state.lamda
+
+        Returns
+        -------
+        state: SMCState,
+            The new state of the tempered SMC algorithm
+         info: SMCInfo,
+            Additional information on the SMC step
+        """
+        n_particles = jax.tree_flatten(state.particles)[0].shape[0]
+
+        def tempered_potential_fn(pytree):
+            return -state.lamda * potential_fn(pytree)
+
+        step_mcmc_kernel = jax.vmap(mcmc_kernel(tempered_potential_fn, mcmc_parameters), in_axes=(0, 0))
```

Comment:
How about?

```python
tempered_potential_fn = lambda position: -state.lmbda * potential_fn(position)
kernel = mcmc_kernel_factory(tempered_potential_fn)
```

Reply:
That depends how much of a PEP purist you want to be. Assigning lambda functions to variables is now advised against as it messes up with the stack trace at debug time.

Reply:
I'm a purist within reason. This specific check is disabled on this repo. Since the codebase is purely functional, disallowing lambda function assignment would make the code harder to read.